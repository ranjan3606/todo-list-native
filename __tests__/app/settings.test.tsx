import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import SettingsScreen from '../../app/(tabs)/settings';

// Mock required modules
jest.mock('@expo/vector-icons', () => ({
  FontAwesome: 'FontAwesome',
}));

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { 
      background: '#fff', 
      text: '#000', 
      tint: '#007AFF',
      lightText: '#666',
      cardBackground: '#f0f0f0',
      border: '#e0e0e0',
    },
    dark: { 
      background: '#000', 
      text: '#fff', 
      tint: '#0A84FF',
      lightText: '#999',
      cardBackground: '#1c1c1e',
      border: '#444',
    },
    primary: '#007AFF',
  }
}));

jest.mock('@/services/theme', () => ({
  useActualColorScheme: () => 'light',
  useThemePreference: () => ['system', jest.fn()],
  saveThemePreference: jest.fn(),
}));

jest.mock('@/services/storage', () => ({
  getSnoozeDuration: () => Promise.resolve(24),
  saveSnoozeDuration: jest.fn(() => Promise.resolve()),
}));

jest.mock('@/hooks/useTags', () => ({
  useTags: () => ({
    tags: {},
    isLoading: false,
    refresh: jest.fn(),
    activeTag: null,
    slideAnim: {},
    handleTagPress: jest.fn(),
  }),
}));

jest.mock('@/components/ThemeSelector', () => ({
  ThemeSelector: jest.fn(() => <React.Fragment>Theme Selector Component</React.Fragment>),
}));

jest.mock('@/components/LanguageSelector', () => ({
  LanguageSelector: jest.fn(() => <React.Fragment>Language Selector Component</React.Fragment>),
}));

jest.mock('@/components/TagForm', () => ({
  TagForm: jest.fn(() => null),
}));

jest.mock('@/components/TagItem', () => ({
  TagItem: jest.fn(() => <React.Fragment>Tag Item Component</React.Fragment>),
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key) => {
      const translations = {
        'settings.title': 'Settings',
        'settings.theme': 'Theme',
        'settings.language': 'Language',
        'settings.snoozeDuration': 'Snooze Duration',
        'settings.defaultSnoozeTime': 'Default Snooze Time',
        'task.hour': 'hour',
        'task.hours': 'hours',
        'settings.snoozeHelpText': 'This is how long tasks will be snoozed when you use the snooze action.',
        'tags.title': 'Tags',
        'tags.addTag': 'Add Tag',
        'tags.loadingTags': 'Loading tags...',
        'tags.noTags': 'No tags available',
        'app.name': 'Todo App',
        'app.version': 'v1.0.0',
      };
      return translations[key] || key;
    }
  })
}));

jest.mock('@/hooks/useStableRefresh', () => ({
  useStableRefresh: jest.fn(),
}));

jest.mock('@/components/PageLayout', () => ({
  PageLayout: jest.fn(({ children, title }) => (
    <React.Fragment>
      <div>HEADER: {title}</div>
      {children}
    </React.Fragment>
  )),
}));

describe('SettingsScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { getByText, queryByText } = render(<SettingsScreen />);
    
    expect(getByText('HEADER: Settings')).toBeTruthy();
    expect(getByText('Theme')).toBeTruthy();
    expect(getByText('Language')).toBeTruthy();
    expect(getByText('Theme Selector Component')).toBeTruthy();
    expect(getByText('Language Selector Component')).toBeTruthy();
    expect(getByText('Snooze Duration')).toBeTruthy();
    expect(getByText('Default Snooze Time')).toBeTruthy();
    expect(getByText('24 hours')).toBeTruthy();
    expect(getByText('Tags')).toBeTruthy();
    expect(getByText('Add Tag')).toBeTruthy();
    expect(getByText('No tags available')).toBeTruthy();
    expect(getByText('Todo App v1.0.0')).toBeTruthy();
  });
  
  it('renders tag list when tags are available', async () => {
    jest.spyOn(require('@/hooks/useTags'), 'useTags').mockReturnValue({
      tags: {
        'Work': ['job', 'project'],
        'Personal': ['home', 'family'],
      },
      isLoading: false,
      refresh: jest.fn(),
      activeTag: null,
      slideAnim: { Work: { current: 0 }, Personal: { current: 0 } },
      handleTagPress: jest.fn(),
    });
    
    // The test should render without errors, but we can't verify the FlatList content
    // because of the mocked TagItem component
    const { toJSON } = render(<SettingsScreen />);
    expect(toJSON()).toBeTruthy();
  });
});