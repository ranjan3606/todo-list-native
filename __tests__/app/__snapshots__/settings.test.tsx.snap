// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsScreen matches snapshot 1`] = `
[
  <Text>
    HEADER: Settings
  </Text>,
  <RCTScrollView
    style={
      {
        "flex": 1,
        "width": "100%",
      }
    }
  >
    <View>
      <View
        style={
          {
            "marginBottom": 24,
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "600",
                "marginBottom": 10,
              },
              {
                "color": "#000",
              },
            ]
          }
        >
          Theme
        </Text>
        <Text>
          Theme Selector Component
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 24,
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "600",
                "marginBottom": 10,
              },
              {
                "color": "#000",
              },
            ]
          }
        >
          Language
        </Text>
        <Text>
          Language Selector Component
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 24,
            "paddingHorizontal": 20,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 18,
                "fontWeight": "600",
                "marginBottom": 10,
              },
              {
                "color": "#000",
              },
            ]
          }
        >
          Snooze Duration
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#f0f0f0",
              "borderRadius": 8,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 8,
              "opacity": 1,
              "padding": 12,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 16,
                },
                {
                  "color": "#000",
                },
              ]
            }
          >
            Default Snooze Time
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "fontSize": 16,
                    "fontWeight": "500",
                    "marginRight": 8,
                  },
                  {
                    "color": "#000",
                  },
                ]
              }
            >
              24
               
              hours
            </Text>
            <FontAwesome
              color="#000"
              name="pencil"
              size={16}
            />
          </View>
        </View>
        <Text
          style={
            [
              {
                "fontSize": 12,
                "fontStyle": "italic",
                "marginTop": 4,
              },
              {
                "color": "#666",
              },
            ]
          }
        >
          This is how long tasks will be snoozed when you use the snooze action.
        </Text>
      </View>
      <View
        style={
          {
            "marginBottom": 24,
            "paddingHorizontal": 20,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 18,
                  "fontWeight": "600",
                  "marginBottom": 10,
                },
                {
                  "color": "#000",
                },
              ]
            }
          >
            Tags
          </Text>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#007AFF",
                "borderRadius": 20,
                "flexDirection": "row",
                "opacity": 1,
                "paddingHorizontal": 12,
                "paddingVertical": 6,
              }
            }
          >
            <FontAwesome
              color="white"
              name="plus"
              size={16}
            />
            <Text
              style={
                {
                  "color": "white",
                  "fontWeight": "500",
                  "marginLeft": 5,
                }
              }
            >
              Add Tag
            </Text>
          </View>
        </View>
        <RCTScrollView
          data={
            [
              "Work",
              "Personal",
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          getItemLayout={[Function]}
          initialNumToRender={5}
          keyExtractor={[Function]}
          maxToRenderPerBatch={5}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={true}
          renderItem={[Function]}
          scrollEnabled={false}
          scrollEventThrottle={0.0001}
          stickyHeaderIndices={[]}
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onFocusCapture={[Function]}
              style={null}
            >
              <Text>
                Tag Item Component
              </Text>
            </View>
            <View
              onFocusCapture={[Function]}
              style={null}
            >
              <Text>
                Tag Item Component
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginBottom": 20,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 14,
              },
              {
                "color": "#666",
              },
            ]
          }
        >
          Todo App
           
          v1.0.0
        </Text>
      </View>
    </View>
  </RCTScrollView>,
]
`;
